bees/prometheusbee/prometheusbee.go

```diff
@@ -0,0 +1,253 @@
+/*
+ *    Copyright (C) 2019 	  CalmBit
+ *                  2014-2019 Christian Muehlhaeuser
+ *
+ *    This program is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU Affero General Public License as published
+ *    by the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *    Authors:
+ *      CalmBit <calmbit@posteo.net>
+ *      Christian Muehlhaeuser <muesli@gmail.com>
+ */
+
+// Package prometheusbee is a bee designed to expose metrics for scraping
+// by Prometheus.
+package prometheusbee
+
+import (
+	"github.com/muesli/beehive/bees"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"net/http"
+)
+
+// PrometheusBee is a bee designed to expose metrics for scraping
+// by Prometheus.
+type PrometheusBee struct {
+	bees.Bee
+
+	addr string
+
+	counterVecName   string
+	gaugeVecName     string
+	histogramVecName string
+	summaryVecName   string
+
+	counter   *prometheus.CounterVec
+	gauge     *prometheus.GaugeVec
+	histogram *prometheus.HistogramVec
+	summary   *prometheus.SummaryVec
+}
+
+// Run executes the Bee's event loop.
+func (mod *PrometheusBee) Run(eventChan chan bees.Event) {
+
+	// Counter vector registration
+
+	mod.counter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: mod.counterVecName,
+			Help: "Collection of all counter variables being iterated by hives",
+		},
+		[]string{"name"},
+	)
+	err := prometheus.Register(mod.counter)
+	if err != nil {
+		mod.LogErrorf("Error registering counter vector: %v", err)
+		return
```

Comment:
We should probably panic for these kind of errors, instead of simply returning. This is a pretty fatal situation, no?

Reply:
Fair enough! I guess I mistakenly thought it wouldn't require a panic. Apologies :)